### ISSUE TITLE
Some API extensions

### ISSUE THREAD

**Issue by stevedonovan** [FascinatedBox/lily#220]
What about `read_all(f: File):ByteString`?  It often makes sense to pull in the whole of a file, if it isn't 'too large'.

`String.find(self, needle: String, *start: Integer):Option[Integer]`. This is useful for moseying through a large string.

`String.size(self: String): Integer` - For consistency with `Hash` and `List`. Can't iterate through a string without this.  (`String.each` needs some thought. I would expect it to iterate over all UTF-8 'characters')

`String.slice(self: String, start,end: Integer): String`. Makes a copy - `end` can be negative a la Python. Empty string if out of bounds?

(Ditto for ByteString)

`ByteString.decode()` could have an optional mode, since that's what people will mostly use (check UTF-8)

I'm happy to submit a PR for some or all of this, if acceptable. Consider them 'serving suggestions' as they say on the bottle of spaghetti sauce.

**Comment by jesserayadkins** [comment]
> File.read_all

Okay.

> String.find

I'm assuming that this would start at a character offset, and would thus need to check for not starting in the middle of some utf-8.

> String.size

The aim for consistency is nice, but I can see a use later on for someone wanting a grapheme count. I'd also like something that makes it more clear what you're getting. String.byte_count and String.grapheme_count, maybe? I need to take a look at what Swift does there.

If the indexes are reversed, that's a good question. An empty String seems okay. 

> String.slice

Hmm...What are we working with in terms of indexes? What happens if the start index is in the middle of a utf-8 chunk, or the end is the same way? At the same time, having slice return an Option...I could see people being annoyed by that.

With ByteString it's more straightforward. But at the same time, ByteString is not very useful right now.

> ByteString.decode

Having more parameters here is fine.


At the moment, it's moderately painful to add new functions because the builtin package has a custom loader and a custom dynaload table (and neither is documented well). I've been meaning to expand dyna_tools.py to allow autogeneration of both, since I need that to add Byte later on. But it isn't impossible.

I'm wary of expanding String too much right now. I'd like to first see utf8proc used for Lily's utf8 handling, and then get used to it in different areas. The current utf-8 handling isn't much, and it's going to need to go sooner or later.

**Comment by stevedonovan** [comment]
Yes, we're still in an area of string-uncertainty.  Ah, I see what you mean about the pain of adding new functions.  Maybe we should just wait for it all settle down then.

`byte_count` is nice & clear - a useful distinction.  

`File.popen(cmd, mode: String): File` would be a very useful addition. I mention this because `File` is a system object and then it needs to understand that `pclose` must be called instead of `fclose`.

I think what's happening is that I want to use Lily to play with real problems. Early adopters can be a pain if you have a carefully worked-out roadmap ;)

**Comment by jesserayadkins** [comment]
It's been a mixed blessing. I like the enthusiasm, and I love getting the PR's. But I don't like that the api isn't done because I know there are people waiting on me for that. But I'll be glad when the api is done and I can see people use Lily to actually perform useful tasks.

I'm hesitant to put `popen` with `File`. If it was put into `sys`, then someone could create an effective sandbox by simply disabling `sys`, and restricting imports/usages. Even if `File` may be a more reasonable place for it to belong. I need some time to think it over more though.

It's less that I have a roadmap planned out...most of this I think over as it comes along.

**Comment by stevedonovan** [comment]
Oops, I forgot about sandboxing. That's a very important consideration, and in fact `File` is very much something you don't want in core then.

**Comment by FascinatedBox** [comment]
I think what's included in this is done now. Slice methods have been added that use byte boundaries for `String` and `ByteString` alike (and with validity checking for utf-8 in `String`). I changed `read_all` from your lily-extras (thanks again, btw :) ) into a more generic `read` in the style of Python (read N bytes, or use -1 to read all). The others have been done as well.

Closing this. Give me a poke/reopen if there's something I missed.
